File: mobile/lib/models/order.dart
import 'package:json_annotation/json_annotation.dart';

part 'order.g.dart';

@JsonSerializable()
class OrderItemVariantSnapshot {
  final String label;
  final int weightGrams;
  final String sku;

  OrderItemVariantSnapshot({
    required this.label,
    required this.weightGrams,
    required this.sku,
  });

  factory OrderItemVariantSnapshot.fromJson(Map<String, dynamic> json) =>
      _$OrderItemVariantSnapshotFromJson(json);
  Map<String, dynamic> toJson() => _$OrderItemVariantSnapshotToJson(this);
}

@JsonSerializable()
class OrderItem {
  final String productId;
  final String variantId;
  final String nameSnapshot;
  final OrderItemVariantSnapshot variantSnapshot;
  final double unitPrice;
  final int qty;
  final double lineTotal;

  OrderItem({
    required this.productId,
    required this.variantId,
    required this.nameSnapshot,
    required this.variantSnapshot,
    required this.unitPrice,
    required this.qty,
    required this.lineTotal,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) =>
      _$OrderItemFromJson(json);
  Map<String, dynamic> toJson() => _$OrderItemToJson(this);
}

@JsonSerializable()
class OrderPricing {
  final double subtotal;
  final double discount;
  final double deliveryFee;
  final double vat;
  final double grandTotal;

  OrderPricing({
    required this.subtotal,
    this.discount = 0,
    required this.deliveryFee,
    required this.vat,
    required this.grandTotal,
  });

  factory OrderPricing.fromJson(Map<String, dynamic> json) =>
      _$OrderPricingFromJson(json);
  Map<String, dynamic> toJson() => _$OrderPricingToJson(this);
}

@JsonSerializable()
class OrderPayment {
  final String method;
  final String status;
  final String? provider;
  final String? transactionId;

  OrderPayment({
    required this.method,
    required this.status,
    this.provider,
    this.transactionId,
  });

  factory OrderPayment.fromJson(Map<String, dynamic> json) =>
      _$OrderPaymentFromJson(json);
  Map<String, dynamic> toJson() => _$OrderPaymentToJson(this);
}

@JsonSerializable()
class OrderFulfillment {
  final String type;
  final Map<String, dynamic>? addressSnapshot;
  final String? notes;

  OrderFulfillment({required this.type, this.addressSnapshot, this.notes});

  factory OrderFulfillment.fromJson(Map<String, dynamic> json) =>
      _$OrderFulfillmentFromJson(json);
  Map<String, dynamic> toJson() => _$OrderFulfillmentToJson(this);
}

@JsonSerializable()
class OrderHistoryEntry {
  final String status;
  final DateTime at;
  final String? byRole;
  final String? byId;
  final String? note;

  OrderHistoryEntry({
    required this.status,
    required this.at,
    this.byRole,
    this.byId,
    this.note,
  });

  factory OrderHistoryEntry.fromJson(Map<String, dynamic> json) =>
      _$OrderHistoryEntryFromJson(json);
  Map<String, dynamic> toJson() => _$OrderHistoryEntryToJson(this);
}

@JsonSerializable()
class Order {
  @JsonKey(name: '_id')
  final String id;
  final String orderNo;
  final String customerId;
  final List<OrderItem> items;
  final OrderPricing pricing;
  final OrderPayment payment;
  final OrderFulfillment fulfillment;
  final String status;
  final String? adminNotes;
  final List<OrderHistoryEntry> history;
  final DateTime createdAt;
  final DateTime updatedAt;

  Order({
    required this.id,
    required this.orderNo,
    required this.customerId,
    required this.items,
    required this.pricing,
    required this.payment,
    required this.fulfillment,
    required this.status,
    this.adminNotes,
    this.history = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);
  Map<String, dynamic> toJson() => _$OrderToJson(this);

  int get totalItems => items.fold(0, (sum, item) => sum + item.qty);

  String get statusDisplay {
    switch (status) {
      case 'PENDING_CONFIRMATION':
        return 'Pending Confirmation';
      case 'CONFIRMED':
        return 'Confirmed';
      case 'PREPARING':
        return 'Preparing';
      case 'OUT_FOR_DELIVERY':
        return 'Out for Delivery';
      case 'DELIVERED':
        return 'Delivered';
      case 'CANCELLED':
        return 'Cancelled';
      case 'REFUNDED':
        return 'Refunded';
      default:
        return status;
    }
  }
}

File: backend/models/Order.js
const mongoose = require('mongoose');

// Order status enum
const ORDER_STATUS = {
  PENDING_CONFIRMATION: 'PENDING_CONFIRMATION',
  CONFIRMED: 'CONFIRMED',
  PREPARING: 'PREPARING',
  OUT_FOR_DELIVERY: 'OUT_FOR_DELIVERY',
  DELIVERED: 'DELIVERED',
  CANCELLED: 'CANCELLED',
  REFUNDED: 'REFUNDED'
};

// Payment status enum
const PAYMENT_STATUS = {
  PENDING: 'PENDING',
  PAID: 'PAID',
  FAILED: 'FAILED',
  REFUNDED: 'REFUNDED'
};

// Payment method enum
const PAYMENT_METHOD = {
  COD: 'COD',
  CARD: 'CARD'
};

// Fulfillment type enum
const FULFILLMENT_TYPE = {
  DELIVERY: 'DELIVERY',
  PICKUP: 'PICKUP'
};

// Order item schema (with snapshots)
const OrderItemSchema = new mongoose.Schema({
  productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
  variantId: { type: mongoose.Schema.Types.ObjectId, required: true },
  
  // Snapshot at purchase time (prices may change later)
  nameSnapshot: { type: String, required: true },
  variantSnapshot: {
    label: { type: String, required: true },      // "250g"
    weightGrams: { type: Number, required: true }, // 250
    sku: { type: String, required: true }
  },
  
  unitPrice: { type: Number, required: true },    // Price per unit at purchase
  qty: { type: Number, required: true, min: 1 },
  lineTotal: { type: Number, required: true }     // unitPrice * qty
}, { _id: true });

// Order history entry
const HistoryEntrySchema = new mongoose.Schema({
  status: { type: String, required: true },
  at: { type: Date, default: Date.now },
  byRole: { type: String },                       // 'CUSTOMER', 'OWNER', 'MANAGER', 'STAFF', 'SYSTEM'
  byId: { type: mongoose.Schema.Types.ObjectId },
  note: { type: String }
}, { _id: true });

const OrderSchema = new mongoose.Schema({
  // Unique order number: CBW-YYYY-######
  orderNo: { type: String, required: true, unique: true },
  
  customerId: { type: mongoose.Schema.Types.ObjectId, ref: 'Customer', required: true },
  
  items: [OrderItemSchema],
  
  // Pricing breakdown (calculated by server)
  pricing: {
    subtotal: { type: Number, required: true },     // Sum of line totals
    discount: { type: Number, default: 0 },         // Coupon/promo discount
    deliveryFee: { type: Number, default: 0 },
    vat: { type: Number, default: 0 },              // 5% VAT if applicable
    grandTotal: { type: Number, required: true }
  },
  
  // Payment info
  payment: {
    method: { 
      type: String, 
      enum: Object.values(PAYMENT_METHOD),
      required: true 
    },
    status: { 
      type: String, 
      enum: Object.values(PAYMENT_STATUS),
      default: PAYMENT_STATUS.PENDING
    },
    provider: { type: String },                     // Future: Stripe, etc.
    transactionId: { type: String }
  },
  
  // Fulfillment info
  fulfillment: {
    type: { 
      type: String, 
      enum: Object.values(FULFILLMENT_TYPE),
      required: true 
    },
    addressSnapshot: {
      name: String,
      phone: String,
      street: String,
      city: String,
      emirate: String,
      building: String,
      apartment: String,
      instructions: String
    },
    notes: { type: String }                         // Customer notes
  },
  
  // Order status
  status: { 
    type: String, 
    enum: Object.values(ORDER_STATUS),
    default: ORDER_STATUS.PENDING_CONFIRMATION
  },
  
  // Admin internal notes
  adminNotes: { type: String },
  
  // Status change history
  history: [HistoryEntrySchema]

}, { timestamps: true });

// Indexes
OrderSchema.index({ orderNo: 1 }, { unique: true });
OrderSchema.index({ customerId: 1 });
OrderSchema.index({ status: 1 });
OrderSchema.index({ createdAt: -1 });
OrderSchema.index({ 'payment.status': 1 });

// Generate order number: CBW-2025-000001
OrderSchema.statics.generateOrderNo = async function() {
  const year = new Date().getFullYear();
  const prefix = `CBW-${year}-`;
  
  // Find the last order of this year
  const lastOrder = await this.findOne({ 
    orderNo: new RegExp(`^${prefix}`) 
  }).sort({ orderNo: -1 });
  
  let nextNum = 1;
  if (lastOrder) {
    const lastNum = parseInt(lastOrder.orderNo.split('-')[2], 10);
    nextNum = lastNum + 1;
  }
  
  return `${prefix}${String(nextNum).padStart(6, '0')}`;
};

// Add history entry helper
OrderSchema.methods.addHistory = function(status, byRole, byId, note) {
  this.history.push({
    status,
    at: new Date(),
    byRole,
    byId,
    note
  });
};

// Export enums for use elsewhere
OrderSchema.statics.ORDER_STATUS = ORDER_STATUS;
OrderSchema.statics.PAYMENT_STATUS = PAYMENT_STATUS;
OrderSchema.statics.PAYMENT_METHOD = PAYMENT_METHOD;
OrderSchema.statics.FULFILLMENT_TYPE = FULFILLMENT_TYPE;

module.exports = mongoose.model('Order', OrderSchema);
